### PR TITLE
[ADD][11.0] purchase_order_general_discount: New Module

### PR DESCRIPTION
**Pull request by chienandalu**
This module allows to set a general discount in a purchase order. This general
discount is applied to each line order.

You can also set a default general discount on suppliers.

cc @Tecnativa

### CODE REVIEW

**Comment by chienandalu** [comment]
There's some issue with travis (https://travis-ci.org/OCA/purchase-workflow/jobs/577689240#L375) Any clue, @moylop260 ?

cc @pedrobaeza

**Comment by rafaelbn** [comment]
IMHO please add/create de default parameter: 

Key | purchase.general_discount_field
-- | --
Value | discount

**Review by rafaelbn — APPROVED** [review]
Functional Test :+1:

**Comment by pedrobaeza** [comment]
Travis issue fixed in https://github.com/OCA/purchase-workflow/commit/10a5377766b61ebec0da5862072701a76a4f12bc. Please rebase.

File: purchase_order_general_discount/data/ir_config_parameter.xml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0"?>
+<odoo>
+    <data noupdate="1">
+
+         <record id="purchase_general_discount_field" model="ir.config_parameter">
+            <field name="key">purchase.general_discount_field</field>
```
Comment (pedrobaeza) [inline purchase_order_general_discount/data/ir_config_parameter.xml]: You must do this at res.company level

File: purchase_order_general_discount/readme/CONFIGURE.rst [hunk 2]
```diff
@@ -0,0 +1,9 @@
+You can set in parameters another discount field to be applied.
+For example, if we had `purchase_triple_dicount`, we could set the general
```
Comment (pedrobaeza) [inline purchase_order_general_discount/readme/CONFIGURE.rst]: s/purchase_triple_dicount/purchase_triple_discount

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Comment by chienandalu** [comment]
@pedrobaeza Changes done

**Comment by chienandalu** [comment]
Ready for final review

**Comment by rafaelbn** [comment]
In my side :+1:

File: purchase_order_general_discount/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+# Copyright 2019 Tecnativa - David Vidal
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+{
+    'name': 'Purchase Order General Discount',
+    'summary': 'General discount per purchase order',
+    'version': '11.0.1.0.0',
+    'category': 'Sales',
```
Comment (pedrobaeza) [inline purchase_order_general_discount/__manifest__.py]: Wrong category

File: purchase_order_general_discount/README.rst [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+**Table of contents**
+
+.. contents::
+   :local:
+
+Configuration
+=============
+
+You can set in settings another discount field to be applied.
+For example, if we had `purchase_triple_discount`, we could set the general
```
Comment (pedrobaeza) [inline purchase_order_general_discount/README.rst:43]: s/had/have

File: purchase_order_general_discount/models/purchase_order.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        return discount_field or 'discount'
+
+    @api.onchange('general_discount')
+    def onchange_general_discount(self):
+        discount_field = self._get_general_discount_field()
+        self.mapped('order_line').update({
+            discount_field: self.general_discount,
+        })
+
+    def action_update_general_discount(self):
```
Comment (pedrobaeza) [inline purchase_order_general_discount/models/purchase_order.py:42]: Why do you need this?

File: purchase_order_general_discount/models/res_company.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+             'discounts will be applied.',
+    )
+
+    @api.model
+    def _get_purchase_discount_fields(self):
+        """Extensible method to add possible discounts. We offer in advance
+           the posibility of using purchase_triple_discount so no bridge
+           module is needed"""
+        discount_fields = [('discount', _('Discount'))]
+        if self.env['ir.module.module'].search([
```
Comment (pedrobaeza) [inline purchase_order_general_discount/models/res_company.py]: Incorrect module search? Any way, it's better to just check in `self.env['purchase.order.line']._fields.keys()` if the fields exist.

File: purchase_order_general_discount/models/res_config_settings.py [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo import fields, models
+
+
+class ResConfigSettings(models.TransientModel):
+    _inherit = 'res.config.settings'
+
+    purchase_general_discount_field = fields.Selection(
+        related='company_id.purchase_general_discount_field',
+        default='discount',
```
Comment (pedrobaeza) [inline purchase_order_general_discount/models/res_config_settings.py:11]: Put `readonly=False` in advanced for v12

File: purchase_order_general_discount/readme/CONFIGURE.rst [hunk 8]
```diff
@@ -0,0 +1,21 @@
+You can set in settings another discount field to be applied.
+For example, if we had `purchase_triple_discount`, we could set the general
+discount in discount3 to be applied after all other discounts.
+
+To do so:
+
+#. Go to *Purchses > Configuration > Settings* and *Purchase Discount Field*
```
Comment (pedrobaeza) [inline purchase_order_general_discount/readme/CONFIGURE.rst]: s/Purchses/Purchases

File: purchase_order_general_discount/readme/USAGE.rst [hunk 9]
```diff
@@ -0,0 +1,10 @@
+To set a partner discount you need to:
```
Comment (pedrobaeza) [inline purchase_order_general_discount/readme/USAGE.rst]: s/partner discount/partner default global purchase discount

File: purchase_order_general_discount/readme/USAGE.rst [hunk 10]
```diff
@@ -0,0 +1,10 @@
...
+
+#. Go to a partner and set the general discount in *Sales & Purchases* tab.
+
+To set a discount in a purchase order you need to:
```
Comment (pedrobaeza) [inline purchase_order_general_discount/readme/USAGE.rst]: s/discount/global discount

File: purchase_order_general_discount/models/purchase_order.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+    _sql_constraints = [
+        ('general_discount_limit', 'CHECK (general_discount <= 100.0)',
+         'Discount must be lower than 100%.'),
+    ]
+
+    @api.onchange('partner_id')
+    def onchange_partner_id(self):
+        super().onchange_partner_id()
+        self.general_discount = self.partner_id.purchase_general_discount
```
Comment (pedrobaeza) [inline purchase_order_general_discount/models/purchase_order.py]:
```suggestion
        self.general_discount = self.partner_id.commercial_partner_id.purchase_general_discount
```

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: purchase_order_general_discount/README.rst (hunk 4 above)
Comment (chienandalu) [reply purchase_order_general_discount/README.rst:43]: It's ok as is conditional. E.g.: 

> If I were a rich man,
> Yubby dibby dibby dibby dibby dibby dibby dum.
> All day long I'd biddy biddy bum.

File: purchase_order_general_discount/models/purchase_order.py (hunk 5 above)
Comment (chienandalu) [reply purchase_order_general_discount/models/purchase_order.py:42]: I use it as wrapper to trigger it from ui

**Comment by chienandalu** [comment]
@pedrobaeza Changes Done

**Comment by rafaelbn** [comment]
@pedrobaeza in my side tested is :+1:  you comments have been attneded, let see if we can merge this one please, thanks!

**Review by ernestotejeda — APPROVED** [review]
Functional and code OK

**Review by pedrobaeza — APPROVED** [review]

**Pull request closed without merging**